### ISSUE TITLE
What is the inspiration for the design of copyist?

### ISSUE THREAD

**Issue by zachvalenta** [cockroachdb/copyist#8]
Just curious to hear more about what inspired the design of copyist, whether another library or CS concept. Never come across a mocking library explained in these terms so I'm interested to delve deeper.

**Comment by andy-kimball** [comment]
The inspiration was the pain of slow, flaky tests. I tried a few different designs at higher levels, such as at the ORM level, before realizing that the recording could be done down at the Go SQL driver level. After several refinement iterations, I got it simplified down to what it is now.

After I'd built `copyist`, another developer at CRL pointed out that there are similar libraries for HTTP, like [https://github.com/dnaeon/go-vcr](https://github.com/dnaeon/go-vcr) (which we have since started using to record and replay Kubernetes calls). So the general idea seems to have been kicking around for some time; I discovered it independently. Credit goes to the Go team for building a nice SQL abstraction layer that allows this kind of hooking.
